src/main/java/com/sapient/model/service/UserService.java

```diff
@@ -1,25 +1,44 @@
 package com.sapient.model.service;
 
+import com.sapient.exception.EmailAlreadyTakenException;
 import com.sapient.exception.UserNotFoundException;
 import com.sapient.exception.UsernameAlreadyTakenException;
 import com.sapient.model.beans.User;
 import com.sapient.model.dao.UserRepository;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
+import at.favre.lib.crypto.bcrypt.BCrypt;
+
+import java.util.Optional;
 
 @Service
 public class UserService {
     @Autowired
     private UserRepository userDao;
 
-    public User createUser(String username, String password, String email) throws UsernameAlreadyTakenException {
+    public User createUser(
+            String username,
+            String password,
+            String email,
+            String firstName,
+            String lastName,
+            Optional<String> phoneNumber)
+            throws UsernameAlreadyTakenException, EmailAlreadyTakenException {
         if (usernameTaken(username)) {
             throw new UsernameAlreadyTakenException();
         }
+        if (emailTaken(email)) {
+            throw new EmailAlreadyTakenException("Email '" + email + "' is already taken.");
+        }
         User user = new User();
         user.setUsername(username);
-        user.setPasswordHash(password);
+        user.setPasswordHash(BCrypt.withDefaults().hashToString(12, password.toCharArray()));
```

Comment:
we should use username as salt

Reply:
Now we concat username and password before hashing.